lib/internal/modules/esm/loader.js

```diff
@@ -822,10 +839,13 @@ class ESMLoader {
 
     const nextResolve = nextHookFactory(chain, meta, validate);
 
-    const resolution = await nextResolve(originalSpecifier, context);
+    const resolution = nextResolve(originalSpecifier, context);
     const { hookErrIdentifier } = meta; // Retrieve the value after all settled
 
-    if (typeof resolution !== 'object') { // [2]
+    if (
+      typeof resolution !== 'object' || // [2]
+      typeof resolution?.then === 'function'
```

Comment:
```suggestion
      resolution === null ||
      typeof resolution.then === 'function'
```

Reply:
Instead of checking here if the returned object is a thenable, I would suggest first try to see if the expected properties are there. If the expected properties are there, no need to check if the returned value has a `then` property imo.

Reply:
I think having non-Promise thenables only _sometimes_ be consumed would be very confusing - if thenables are respected it seems like checking for .then should be done always

Reply:
No, I mean for `resolve` hook, which is now sync, we never want to consume the thenable. However, if `resolve` returns a thenable, that would be a sign that the user is using an async function for the `resolve` hook so it would be helpful to warn them about this.
My proposal is: don't check if there's a `then` property until the other checks as failed (so when we're on the error path anyway).

Reply:
Perfect, that makes sense then